### PR TITLE
REF/MAINT: deprecate sjoin op param in favor of predicate

### LINKED ISSUE #1525: Consistency of predicate argument
I have realised that with changes to sindex in 0.8.0 we have introduced inconsistency in terminology. While in `sjoin`, we have an attribute `op`, in `sindex.query` we use `predicate`. Both mean the same thing and are used for the same purpose. 

Not sure what the best option would be, but I quite like literal `predicate` as I always found `op` a bit cryptic (btw, what the `o` stands for? overlay? as overlay predicate?). One way could be to have both as synonyms or to have only `predicate` but silently accept `op` as well (until some deprecation or just keep that). But even I was confused by the current situation while trying to use `op` in query.

cc @adriangb

### PR DESCRIPTION
**Pull request by adriangb**
Closes #1525

### CODE REVIEW

**Review by brendan-ward — CHANGES_REQUESTED** [review]
@adriangb thanks for working on this, I think `predicate` is much more clear now in this context than `op`.  👍 

Since this deprecates a parameter in a top-level API, it might be good to target this to 0.9 instead of next 0.8.x release (only effect here being that we wait a bit to merge).

Will also need slight update in upcoming `sjoin` documentation @martinfleis

**Comment by adriangb** [comment]
Where are we with this @martinfleis ? Do you want me to resolve conflicts and fix anything else?

**Review by martinfleis — COMMENTED** [review]
Ah, this has slipped from my mind. Should've been in 0.9.0.

Can you just change `op` to `predicate` in the user guide? https://geopandas.readthedocs.io/en/latest/docs/user_guide/mergingdata.html#spatial-joins

File: geopandas/tools/sjoin.py [hunk 1]
```diff
@@ -21,19 +27,32 @@ def sjoin(
...
     lsuffix : string, default 'left'
         Suffix to apply to overlapping column names (left GeoDataFrame).
     rsuffix : string, default 'right'
         Suffix to apply to overlapping column names (right GeoDataFrame).
+    **kwargs: dict
+        For handling of deprecated `op` param, which will be replaced by
+        `predicate`.
     """
+    if "op" in kwargs:
+        warnings.warn(
```
[the remark this answers is not in the document]
Comment (adriangb) [reply geopandas/tools/sjoin.py]: Swapped

**Comment by adriangb** [comment]
Hey @martinfleis, I made the doc edits and swapped the `UserWarning -> FutureWarning`.

I also realized that there's 2 edge cases that we weren't checking for:
- Both `op` and `predicate` are specified. I added a `UserWarning` for this that is only emitted if `predicate!="intersects"` (i.e. not the default) _and_ `op!=predicate`. This also includes the deprecation message. Let me know if you'd rather this be a `FutureWarning`.
- Extra `**kwargs` are specified. This will now raise a `ValueError` that lists out the extra kwargs.

**Review by martinfleis — APPROVED** [review]
Thanks @adriangb!

**Comment by adriangb** [comment]
@brendan-ward bumping again, hoping this doesn't get left out of the next release

**Comment by martinfleis** [comment]
@adriangb can you make sure this is mergeable (no conflicts, green CI)?

**Comment by adriangb** [comment]
yeah let me resolve conflicts

**Comment by adriangb** [comment]
Conflicts resolved. CI failures look like the same Matplotlib issue as #2066

File: geopandas/tools/sjoin.py [hunk 2]
```diff
@@ -84,6 +94,33 @@ def sjoin(
...
+                f' (got `predicate="{predicate}"`'
+                f' in combination with `op` (got `op="{op}"`).'
+                " The value of `predicate` will be overriden by the value of `op`,"
+                " , which may result in unexpected behavior."
+                f"\n{deprectation_message}"
+            )
+            warnings.warn(override_message, UserWarning, stacklevel=4)
+        else:
+            warnings.warn(deprectation_message, FutureWarning, stacklevel=4)
+    if kwargs:
```
Comment (brendan-ward) [inline geopandas/tools/sjoin.py:116]: Would be good to model this off other functions, e.g.,  

```
>>> df.contains(df.geometry, foo='bar')
...
TypeError: contains() got an unexpected keyword argument 'foo'
```

**Review by brendan-ward — COMMENTED** [review]
@adriangb thanks for working on this!

A few minor suggestions, but then otherwise this looks good to me.  The CI failures seem unrelated.

File: geopandas/tools/sjoin.py [hunk 3]
```diff
@@ -24,15 +30,19 @@ def sjoin(
...
         You can check the valid values in left_df or right_df as
         ``left_df.sindex.valid_query_predicates`` or
         ``right_df.sindex.valid_query_predicates``
+        Replaces deprecated ``op`` parameter.
     lsuffix : string, default 'left'
         Suffix to apply to overlapping column names (left GeoDataFrame).
     rsuffix : string, default 'right'
         Suffix to apply to overlapping column names (right GeoDataFrame).
+    **kwargs: dict
+        For handling of deprecated `op` param, which will be replaced by
```
[the remark this answers is not in the document]
Comment (adriangb) [reply geopandas/tools/sjoin.py:44]: ok

File: geopandas/tools/sjoin.py (hunk 2 above)
Comment (adriangb) [reply geopandas/tools/sjoin.py:116]: makes sense, simpler as well

File: geopandas/tools/sjoin.py [hunk 4]
```diff
@@ -84,6 +94,33 @@ def sjoin(
...
+                "A non-default value for `predicate` was passed"
+                f' (got `predicate="{predicate}"`'
+                f' in combination with `op` (got `op="{op}"`).'
+                " The value of `predicate` will be overriden by the value of `op`,"
+                " , which may result in unexpected behavior."
+                f"\n{deprectation_message}"
+            )
+            warnings.warn(override_message, UserWarning, stacklevel=4)
+        else:
+            warnings.warn(deprectation_message, FutureWarning, stacklevel=4)
```
[the remark this answers is not in the document]
Comment (adriangb) [reply geopandas/tools/sjoin.py:115]: great catch!

[the remark this answers is not in the document]
Comment (adriangb) [reply geopandas/tools/sjoin.py:115]: I added:

```python
@pytest.mark.parametrize("predicate_kw", ["predicate", "op"])
```

To the `test_inner` test. I think this is the easiest way to catch this bug.

**Comment by adriangb** [comment]
All tests are passing now w/ master merged in, and I believe I've addressed all of the feedback. Please ping me if there's anything else needed @brendan-ward @martinfleis

**Comment by adriangb** [comment]
@martinfleis is there anything blocking a merge?

**Comment by martinfleis** [comment]
> @martinfleis is there anything blocking a merge?

I'd like to have another review approval here.

**Comment by adriangb** [comment]
Got it. Happy to wait 😄

**Comment by martinfleis** [comment]
Here we go :).

Thanks @adriangb!

**Pull request merged**